data/txHandler.go

```diff
@@ -282,41 +309,135 @@ func (handler *TxHandler) asyncVerifySignature(arg interface{}) interface{} {
 		// we failed to write to the output queue, since the queue was full.
 		// adding the metric here allows us to monitor how frequently it happens.
 		transactionMessagesDroppedFromPool.Inc(nil)
+
+		// delete from duplicate caches to give it chance to be re-submitted
+		// this relatively rare operation and implementation is expensive (requires re-hashing)
+		handler.deleteFromCaches(tx.rawmsg.Data, tx.unverifiedTxGroup)
+
 	}
 	return nil
 }
 
+func (handler *TxHandler) deleteFromCaches(msg []byte, unverifiedTxGroup []transactions.SignedTxn) {
+	var result []byte
+	for i := range unverifiedTxGroup {
+		encodeBuf := unverifiedTxGroup[i].MarshalMsg(nil)
+		result = append(result, encodeBuf...)
+	}
+	d := crypto.Hash(result)
+	handler.txCanonicalCache.Delete(&d)
+	handler.msgCache.Delete(msg)
+}
+
+// dedupCanonical checks if the transaction group has been seen before after reencoding to canonical representation.
+// returns a key used for insertion if the group was not found.
+func (handler *TxHandler) dedupCanonical(ntx int, unverifiedTxGroup []transactions.SignedTxn, consumed int) (*crypto.Digest, bool) {
```

Comment:
nit: naming the return variables will make it easier to get their semantics. 
```suggestion
func (handler *TxHandler) dedupCanonical(ntx int, unverifiedTxGroup []transactions.SignedTxn, consumed int) (key *crypto.Digest, isDuplicate bool) {
```